### ISSUE TITLE
KeyError: "'friendlycountryname' is not an existing key"

### ISSUE THREAD

**Issue by mmascolino** [cannatag/ldap3#770]
I recently upgraded my ldap3 package from 2.6 to 2.6.1 and I am now experiencing this error on any query:

```
Traceback (most recent call last):
  File "exportDirectorySearch.py", line 50, in <module>
    main(args.filter, args.basedn, args.attrs, args.outfile)
  File "exportDirectorySearch.py", line 10, in main
    ed = EnterpriseDirectory.EDSearcher(usr, pwd)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\EnterpriseDirectory.py", line 13, in __init__
    self.conn = Connection(server, username, password, auto_bind=True)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\connection.py", line 326, in __init__
    self.do_auto_bind()
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\connection.py", line 343, in do_auto_bind
    self.bind(read_server_info=True)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\connection.py", line 613, in bind
    self.refresh_server_info()
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\connection.py", line 1348, in refresh_server_info
    self.server.get_info_from_server(self)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\server.py", line 449, in get_info_from_server
    self._get_schema_info(connection)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\core\server.py", line 425, in _get_schema_info
    self._schema_info = SchemaInfo(schema_entry, connection.response[0]['attributes'], connection.response[0]['raw_attributes']) if result else None
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\protocol\rfc4512.py", line 298, in __init__
    self.attribute_types = AttributeTypeInfo.from_definition(attributes.pop('attributeTypes', []))
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\protocol\rfc4512.py", line 543, in from_definition
    ret_dict.set_alias(object_def.name[0], object_def.name[1:] + [object_def.oid], ignore_duplicates=True)
  File "C:\Users\mark\Downloads\pyenterprisedirectory-master\.venv\lib\site-packages\ldap3\utils\ciDict.py", line 172, in set_alias
    raise KeyError('\'' + str(ci_key) + '\' is not an existing key')
KeyError: "'friendlycountryname' is not an existing key"
```

**Comment by cannatag** [comment]
It seems that the server your connecting to is missing information on the friendlycountryname attribute (or its “co” alias, look at https://ldapwiki.com/wiki/FriendlyCountryName). 

Can you try to printout the server schema to see the attribute definition?

Thank,
Giovanni

**Comment by mmascolino** [comment]
I can check.  The odd thing is that with v2.6 of ldap3, I can successfully return the co or friendlyCountryName attribute for entries.

**Comment by mmascolino** [comment]
Using an older version of ldap3, `server.schema` contains:
           `"( 0.9.2342.19200300.100.1.43 NAME ( 'co' 'friendlycountryname' ) DESC 'Standard LDAP attribute type' SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 X-ORIGIN 'RFC 1274' )",`

**Comment by cannatag** [comment]
Hi, didn't you get the same schema from the server with 2.6.1?

**Comment by mmascolino** [comment]
I would assume so but the mere act of binding to the server with 2.6.1 throws the exception so I am not sure how I would check i.e.
```

server = Server('myserver.example.com', use_ssl=True, get_info=ALL)
conn = Connection(server, myuser, mypass, auto_bind=True)

```
throws on the Connection constructor.

**Comment by cannatag** [comment]
What LDAP server are you trying to connect to?

**Comment by mmascolino** [comment]
RadiantLogic

**Comment by ocaballeror** [comment]
Same problem here with my corporate LDAP, except my missing key is `streetaddress`.

Similarly, downgrading to 2.6.0 fixes the issue for me.

**Comment by cannatag** [comment]
Hi, I'm not able to replicate this. If I try to parse the attribute description you sent I get the correct attribute definition:

```
Attribute type: 0.9.2342.19200300.100.1.43
  Short name: co, friendlycountryname
  Description: Standard LDAP attribute type
  Single value: False
  Syntax: 1.3.6.1.4.1.1466.115.121.1.15 [('1.3.6.1.4.1.1466.115.121.1.15', 'LDAP_SYNTAX', 'Directory String', 'RFC4517')]
  Mandatory in: friendlyCountry
  Extensions:
    X-ORIGIN: RFC 1274
  OidInfo: ('0.9.2342.19200300.100.1.43', 'ATTRIBUTE_TYPE', ['co', 'friendlyCountryName'], 'RFC4524')```

Could you try the following with 2.6.1:
```
from ldap3 import Connection, Server, NONE, BASE

s = Server('localhost', get_info=NONE)
c = Connection(s, 'my_user', 'my_password')
c.bind()
c.search('cn=Subschema', '(objectClass=*)', BASE, attributes=['subschemaSubentry'], get_operational_attributes=True)
c.response_to_file('myldap.schema')
c.unbind()```

This reads the schema without parsing and write it to the myldap.schema file. Then send the output to me. It should not contain any data from your LDAP server. Thanks.

**Comment by mmascolino** [comment]
[myldap-2.6.1.schema.txt](https://github.com/cannatag/ldap3/files/4118881/myldap-2.6.1.schema.txt)

I've never queried for schema before so check my work...I had to change the base dn to cn=Schema to get any results.  Nothing was returned for subschemaSubentry but I don't know if that is expected.

**Comment by flash3780** [comment]
I can also confirm the problem with 2.6.1 when connecting to my corporate LDAP. My missing key is streetaddress.

Prior versions worked okay.

**Comment by cannatag** [comment]
> 
> 
> [myldap-2.6.1.schema.txt](https://github.com/cannatag/ldap3/files/4118881/myldap-2.6.1.schema.txt)
> 
> I've never queried for schema before so check my work...I had to change the base dn to cn=Schema to get any results. Nothing was returned for subschemaSubentry but I don't know if that is expected.

It seems that the friendlycountryname attribute is defined twice in the schema:
```
"( 0.9.2342.19200300.100.1.43 NAME ( 'co' 'friendlycountryname' ) DESC 'Standard LDAP attribute type' SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 X-ORIGIN 'RFC 1274' )
```
and
```
"( 0.9.2342.19200300.99.1.8 NAME 'friendlyCountryName' EQUALITY caseIgnoreMatch SUBSTR caseIgnoreSubstringsMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.15{256} 
```

Attributes names should be unique in the schema. This should be related to the following fix in 2.6.1:
```
    - Schema info now uses CaseInsensitiveWithAlias dict as default so object and attributes can also be referentiated with OID (thanks ahoffm11)
```

I'm not sure of how to manage this. Any idea?

**Comment by cannatag** [comment]
It would be helpful if you could provide an access to a server that shows the error. A docker image with the server running and a script that show the error should be a better solution.

**Comment by cannatag** [comment]
Didn't notice that the attribute name is cased in different ways(all lowercased in the former and camelcased in the latter). So in the previous versions these were 2 different attributes, but the LDAP protocol states that the attribute names are not case sensitive.

A quick and dirty solution is to set the "CASE_INSENSITIVE_ATTRIBUTE_NAMES" parameter to False with

```
from ldap3 import set_config_parameter
set_config_parameter('CASE_INSENSITIVE_ATTRIBUTE_NAMES', False)
```

This should revert the behaviour if the ldap3 library to the previous version.

Let me know if this solves this issue.

Thanks,
Giovanni

**Comment by mmascolino** [comment]
That unfortunately did not resolve the issue.  I still get the same error.

**Comment by cannatag** [comment]
Can you try the code in dev? It should work. Now the second OID is added as an alias to to first one.

Thanks,
Giovanni

**Comment by mmascolino** [comment]
An error in setting up the alias:

```
(.venv) PS C:\Users\mascolino.mm\Documents\DevWork\pyenterprisedirectory> python .\exportDirectorySearch.py "(cn=masco*)"
Traceback (most recent call last):
  File ".\exportDirectorySearch.py", line 50, in <module>
    main(args.filter, args.basedn, args.attrs, args.outfile)
  File ".\exportDirectorySearch.py", line 10, in main
    ed = EnterpriseDirectory.EDSearcher(usr, pwd)
  File "C:\Users\mascolino.mm\Documents\DevWork\pyenterprisedirectory\EnterpriseDirectory.py", line 14, in __init__
    self.conn = Connection(server, username, password, auto_bind=True)
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\connection.py", line 326, in __init__
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\connection.py", line 343, in do_auto_bind
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\connection.py", line 613, in bind
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\connection.py", line 1348, in refresh_server_info
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\server.py", line 449, in get_info_from_server
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\core\server.py", line 425, in _get_schema_info
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\protocol\rfc4512.py", line 298, in __init__
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\protocol\rfc4512.py", line 543, in from_definition
  File "C:\Users\mascolino.mm\Documents\DevWork\ldap3\.venv\lib\site-packages\ldap3-2.6.1-py3.8.egg\ldap3\utils\ciDict.py", line 172, in set_alias
KeyError: "'friendlycountryname' is not an existing key"
```

**Comment by cannatag** [comment]
Are you sure you're running the "dev" branch? It should be 2.7 not 2.6.1.

You can get the dev branch with:

```
git clone -b dev https://github.com/cannatag/ldap3.git
```
You don't need to install the package, just move to the ldap3 folder and run your python program.

**Comment by mmascolino** [comment]
my mistake...I will give it a try later.

**Comment by mmascolino** [comment]
dev branch works perfectly with or without the `set_config_parameter('CASE_INSENSITIVE_ATTRIBUTE_NAMES', False)`

thank you.

**Comment by cannatag** [comment]
Fixed in ldap3 v2.7.
